### ISSUE TITLE
-n flag not behaving properly, bug?

### ISSUE THREAD

**Issue by toczydlowski** [marcelm/cutadapt#214]
Hello,

I have single-end 100bp GBS data (from Illumina HiSeq) and want to trim the restriction enzyme cut site sequence off the 5' end, and then in the cases where my DNA fragment was (sadly) shorter than 100bp (the length I sequenced), I want to trim the portion of the 3' common adapter that got sequenced.  I know in the test file I was working with that the cut site sequence occurred at the beginning of every read.  The common adapter sequence occurs in about 40-50% of my reads (the other 50-60% of the reads were at least 100bp in length, so the adapter sequence was not sequenced).  See attached picture for visual example of my reads.

RUN1
When I run the below (only 5' cut site sequence), the number of times the adapter is trimmed exactly equals  the number of reads.  This is correct/what I would expect.
-> cutadapt -g ^CWGC -o /Users/(...)/sample1-trimmed.fq  /Users/(...)/sample1.fq 
reads processed: 2,341,093
times trimmed: 2,341,093

RUN2
When I run the below (only 3' common adapter), the number of times the adapter is trimmed is what I would expect (from searching for the adapter sequence using textwrangler).
-> cutadapt -a AGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTG -o /Users/(...)/sample1-trimmed.fq  /Users/(...)/sample1.fq
reads processed: 2,341,093
times trimmed: 1,137,059

Here's the issue:
RUN3
When I run the below (both 5' and 3' adapters included, and n flag added), the 5' cut site sequence is cut more times than the number of sequences I have (concerning/not correct), and the 3' common adapter sequence is cut less.
-> cutadapt -g ^CWGC -a AGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTG -n 2 -o /Users/(...)/sample1-trimmed.fq  /Users/(...)/sample1.fq
reads processed: 2,341,093
1st adapter trimmed: 2,348,770 (should be 2,341,093, from RUN1)
2nd adapter trimmed: 1,136,947 (should be 1,137,059, from RUN2)

In summary - If I run RUN1 and RUN2 separately on the same input, I get the correct number of trims.  If I run RUN1, input RUN1 output as input into RUN2, and run RUN2, I also get the correct number of trims (and the output I actually want, with both adapters trimmed from the reads).  But, if I run the two adapters in the same line of code with -n flag, I get different results.  As far as I understand, I am trying to accomplish what is described in the cutadapt manual in the attached screenshot.

Thanks for any insights!
Rachel
![example of my reads](https://cloud.githubusercontent.com/assets/23200404/20317483/df3096ae-ab2b-11e6-8fc1-c5b82086a9df.png)
![cutadapt manual screenshot](https://cloud.githubusercontent.com/assets/23200404/20317487/e28464c0-ab2b-11e6-9794-7d8370adbc21.png)

**Comment by marcelm** [comment]
Hi, I’ll have a look at your request in more detail tomorrow, but my first guess is that you [perhaps should use the new "linked adapters" option](https://cutadapt.readthedocs.io/en/stable/guide.html#linked-adapters) available since cutadapt 1.10. That section in the documentation is outdated and I will fix it.

**Comment by toczydlowski** [comment]
Hello,


Thanks for checking into this.  I was/am using 1.11.


I am not sure if linked adapters is what I want - I'll look for the updated documentation on this.  The reason I say that is that the 5' adapter will always be present, but the 3' adapter is only sometimes present.  If the linked adapter flag/mode only cuts if it finds the 5' and 3' adapters, then that wouldn't work for me.  If this is not the case though, then yes, perhaps linked adapters would address this.


Currently I am just running the code in two steps - not efficient but gets the job done.


Let me know what you find.


Thanks,

Rachel


Rachel Toczydlowski

PhD Candidate
UW Madison Botany


________________________________

**Comment by marcelm** [comment]
I’ve now read through your request in more detail. Yes, you should use linked adapter option, it was designed for exactly this case. It works the way you want: The 5' adapter must exist, but the 3' adapter is optional.

As an explanation for why the `-n` option does not work: It does not distinguish between the adapters. In every round, it just looks for the best matching adapter and removes it. So there is nothing that prevents it from removing the 5' adapter twice or the 3' adapter twice, which is why you get the 5' adapter trimmed a lot more than you would expect. The solution until before cutadapt 1.10 was to run cutadapt twice, as you have done. It is actually not that inefficient if you use a pipeline to send the output of one cutadapt process directly into the next.

Since you use version 1.11, you can just use the linked adapters. As I said, your use case is exactly what they are designed for.

I will improve the documentation a bit to make all of this clear.

**Comment by marcelm** [comment]
I had actually updated that part of the documentation already, but it is only visible [in the development version](https://cutadapt.readthedocs.io/en/latest/guide.html#trimming-more-than-one-adapter-from-each-read) of the documentation page. I have now also updated [the section on linked adapters](https://cutadapt.readthedocs.io/en/latest/guide.html#linked-adapters).

Please close this issue if you are satisfied.
